brat/Brat/Checker.hs

```diff
@@ -485,6 +485,70 @@ check' (Simple tm) ((), ((hungry, ty):unders)) = do
                                      R0 (RPr ("value", vty) R0)
       wire (dangling, vty, hungry)
       pure (((), ()), ((), unders))
+check' FanOut ((p, ty):overs, ()) = do
+  ty <- eval S0 (binderToValue ?my ty)
+  case ty of
+    TVec elTy n
+      | VNum n <- n
+      , Just n <- numValIsConstant n ->
+          if n < 0
+          then err (InternalError $ "Vector of negative length (" ++ show n ++ ")")
+          else do
+            wires <- fanoutNodes ?my n (p, valueToBinder ?my ty) elTy
+            pure (((), wires), (overs, ()))
+      | otherwise -> typeErr $ "Can't fanout a Vec with non-constant length: " ++ show n
+    _ -> typeErr "Fanout ([/\\]) only applies to Vec"
+ where
+  fanoutNodes :: Modey m -> Integer -> (Src, BinderType m) -> Val Z -> Checking [(Src, BinderType m)]
+  fanoutNodes _ 0 _ _ = pure []
+  fanoutNodes my n (dangling, ty) elTy = do
+    (_, [(hungry, _)], [danglingHead, danglingTail], _) <- anext "fanoutNodes" (Selector (plain "cons")) (S0, Some (Zy :* S0))
+      (RPr ("value", binderToValue my ty) R0)
+      ((RPr ("head", elTy) (RPr ("tail", TVec elTy (VNum (nConstant (n - 1)))) R0)) :: Ro m Z Z)
+    -- Wire the input into the selector node
+    wire (dangling, binderToValue my ty, hungry)
+    (danglingHead:) <$> fanoutNodes my (n - 1) danglingTail elTy
+
+check' FanIn (overs, ((tgt, ty):unders)) = do
+  ty <- eval S0 (binderToValue ?my ty)
+  case ty of
+    TVec elTy n
+      | VNum n <- n
+      , Just n <- numValIsConstant n ->
+          if n < 0
+          then err (InternalError $ "Vector of negative length (" ++ show n ++ ")")
+          else faninNodes ?my n (tgt, valueToBinder ?my ty) elTy overs >>= \case
+            Just overs -> pure (((), ()), (overs, unders))
+            Nothing -> typeErr ("Not enough inputs to make a vector of size " ++ show n)
```

Comment:
Why not move this error case into `faninNodes`?

Reply:
in `faninNodes` we don't have the total size of the vector, just the size of the vector that we have yet to make.
So, this way the off-by-one error [here](https://github.com/CQCL/brat/pull/36/files/5ed5e9c891f6538e4f424d47b81a996ae63a1668#diff-469a8813aeef1b7b613c7adaa41be27f7d737590fff4deb7c70e8f762bfbb900) can be "Not enough inputs to make vector of size 3" rather than "Not enough inputs to make vector of size 1"

Reply:
Of course we could give faninNodes an extra argument for error reporting, I just thought this would be nicer

Reply:
Ok, makes sense 👍